core/utils/cuckoo_map.h

```diff
@@ -0,0 +1,331 @@
+/* Streamlined hash table implementation, with emphasis on lookup performance.
+ * Key and value sizes are fixed. Lookup is thread-safe, but update is not. */
+
+#ifndef BESS_UTILS_CUCKOOMAP_H_
+#define BESS_UTILS_CUCKOOMAP_H_
+
+#include <glog/logging.h>
+
+#include <algorithm>
+#include <limits>
+#include <stack>
+#include <utility>
+#include <vector>
+
+#include "../mem_alloc.h"
+#include "common.h"
+
+namespace bess {
+namespace utils {
+
+// Hash function. Return uint32_t as the hash code.
+template <typename K, typename V>
+using HashFunc = uint32_t (*)(const K& key, uint32_t init_val);
+
+// Compare function. Return true if the lhs and rhs are identical.
+template <typename K, typename V>
+using CmpFunc = bool (*)(const K& lhs, const K& rhs);
+
+// Cuckoo HashMap implementation
+template <typename K, typename V, HashFunc<K, V> H, CmpFunc<K, V> C>
+class CuckooMap {
+ public:
+  typedef std::pair<K, V> Entry;
+
+  CuckooMap()
+      : bucket_mask_(kInitNumBucket - 1),
+        cnt_(0),
+        buckets_(nullptr),
+        entries_(nullptr),
+        num_entries_(kInitNumEntries),
+        free_entry_indices_() {
+    buckets_ = reinterpret_cast<Bucket*>(
+        mem_alloc((bucket_mask_ + 1) * sizeof(Bucket)));
```

Comment:
It's generally a bad idea to use a non c++ allocator for a c++ object that has a defined constructor.  One alternative is to do placement new:

https://isocpp.org/wiki/faq/dtors#placement-new

However you might instead just use new [].